### ISSUE TITLE
[firebase_messaging] onMessage/onResume is called twice on Android/iOS

### ISSUE THREAD

**Issue by woprandi** [FirebaseExtended/flutterfire#1669]
**Describe the bug**
The `onMessage` callback is always called twice 

**To Reproduce**

1. Setup the plugin, do a `print` statement in the `onMessage` callback
2. Keep the app in the foreground
3. Send a FCM message
4. 2 print on the console

**Expected behavior**
Only one print on the console

**Additional context**
I don' t have the same issue with `onBackgroundMessage`. It's called once as expected

**Comment by dhakadv06** [comment]
Facing same issue. onResume and onLaunch works fine but onMessage triggers twice.

**Comment by brian2694** [comment]
Facing same issue.!

**Comment by hoahuathien** [comment]
Facing the same problem.

**Comment by rodrigolmacedo** [comment]
Same issue here, its is boring, in other app with other version the bug dont occours T_T

**Comment by scognito** [comment]
I confirm (firebase messaging 6.0.9, and flutter 1.12.13+hotfix.5)
but it happens only on Android, not iOS.

**Comment by eyoeldefare** [comment]
Only on the android, it seems to me.  Its easy to handle programmatically but not sure if there's a clean way to deal with it.

**Comment by CoderJava** [comment]
Facing same issue.

**Comment by woprandi** [comment]
I confirm it's only on Android

**Comment by woprandi** [comment]
onResume is also called twice

**Comment by sergitovar91** [comment]
Here the same issue, I get duplicate notifications on android. Here you will find all my details. It is the default flutter project with firebase firebase_messaging: ^6.0.9

[✓] Flutter (Channel master, v1.14.2-pre.51, on Mac OS X 10.15.2 19C57, locale en-ES)
    • Flutter version 1.14.2-pre.51
    • Framework revision ec1044a877 (4 days ago), 2020-01-17 19:19:33 -0800
    • Engine revision ecc74c966d
    • Dart version 2.8.0 (build 2.8.0-dev.3.0 f96c48307d)

[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at /Users/sergio/Library/Android/sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-29, build-tools 29.0.2
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b49-5587405)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.3.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.3.1, Build version 11C504
    • CocoaPods version 1.7.5

[✓] Android Studio (version 3.5)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 42.1.1
    • Dart plugin version 191.8593
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b49-5587405)

[✓] Connected device (1 available)
    • Android SDK built for x86 • emulator-5554 • android-x86 • Android 10 (API 29) (emulator)

[code.txt](https://github.com/FirebaseExtended/flutterfire/files/4093495/code.txt)

**Comment by janosdupai** [comment]
same issue on iOS. onMessage called twice with different message ID, so there could be the issue on server side.
I double checked that my php call run only once, but somehow I received 2 message. In both cases: foreground and background too....

**Comment by EverestBoy** [comment]
Same here, on Message is called twice, flutter team are you listening?

**Comment by woprandi** [comment]
@janosdupai I don't have this issue on iOS. I tested again today

**Comment by janosdupai** [comment]
@woprandi 
I use Firebase Cloud Message for a Flutter app.

I try to trigger a push notification with php code. It give me a success: 1 result, but on my device I always receive 2 message with the same title&body but different message ID (which is provided by google firebase).

I double checked, that I trigger this function only once on a server side.

my code on the server
```

    $url = 'https://fcm.googleapis.com/fcm/send';
     $msg = array
          (
                'body'  => $message,
                'title' => $title,
                        'badge' => 1,/*Default sound*/
                        'sound' => 'default',
          );
    $fields = array
            (
                'registration_ids' => $id,
                'notification'  => $msg
            );

    $fields = json_encode ( $fields );

    $apiKey = 'XXXXXXX'; //IOS

    $headers = array (
            'Authorization: key='.$apiKey,
            'Content-Type: application/json'
    );

    $ch = curl_init ();
    curl_setopt ( $ch, CURLOPT_URL, $url );
    curl_setopt ( $ch, CURLOPT_POST, true );
    curl_setopt ( $ch, CURLOPT_HTTPHEADER, $headers );
    curl_setopt ( $ch, CURLOPT_RETURNTRANSFER, true );
    curl_setopt ( $ch, CURLOPT_SSL_VERIFYPEER, false );
    curl_setopt ( $ch, CURLOPT_POSTFIELDS, $fields );

    $result = curl_exec ( $ch );
    curl_close ( $ch );
        echo $result;
```
I tried to trigger this from Terminal, which worked fine, and sent only 1 message. My terminal code:

DATA='{"notification": {"body": "this is a body","title": "this is a title"}, "priority": "high", "data": {"click_action": "FLUTTER_NOTIFICATION_CLICK", "id": "1", "status": "done"}, "to": "MYDEVICETOKEN"}'
curl https://fcm.googleapis.com/fcm/send -H "Content-Type:application/json" -X POST -d "$DATA" -H "Authorization: key=XXXXXXXXX"

**Comment by woprandi** [comment]
@janosdupai Can you check the FCM response via `curl -i` ?

**Comment by janosdupai** [comment]
@woprandi 

{"multicast_id":8502274311835153380,"success":1,"failure":0,"canonical_ids":0,"results":[{"message_id":"1579683773320267"}]}

**Comment by CoderJava** [comment]
I'm test used terminal but the result it's same. OnMessage is called twice. I'm testing on Android device.

**Comment by scognito** [comment]
> 
> 
> @woprandi
> 
> {"multicast_id":8502274311835153380,"success":1,"failure":0,"canonical_ids":0,"results":[{"message_id":"1579683773320267"}]}
> 
> I receive this message_id in flutter app, but I receive another one in a small delay (with same message title and body)....
> 
> on flutter console
> 
> flutter: {status: done, id: 1, click_action: FLUTTER_NOTIFICATION_CLICK, google.c.a.e: 1, gcm.message_id: 1579683773320267, aps: {alert: {title: This is title #1, body: here is a message. message}, sound: default}}
> flutter: {status: done, id: 1, click_action: FLUTTER_NOTIFICATION_CLICK, google.c.a.e: 1, gcm.message_id: 1579685542411682, aps: {alert: {title: This is title #1, body: here is a message. message}, sound: default}}

I receive it two times even from terminal

**Comment by sudip92** [comment]
Facing same issue with firebase_messaging: ^6.0.9. onMessage and onResume call multiple times

**Comment by eyoeldefare** [comment]
> Same here, on Message is called twice, flutter team are you listening?

Can you imagine how many issues there are under this package? You need to be more appreciative of their craft bro. They can't solve everything overnight, give them some time.

**Comment by woprandi** [comment]
After fast investigation, the method `onAttachedToEngine` of the class `FirebaseMessagingPlugin`, which call `registerReceiver`, is called twice. That explain the double call.
I will try to continue to investigate but I'm not an expert

**Comment by woprandi** [comment]
A working workaround :

```
class MainActivity: FlutterActivity() {
   /*  override fun configureFlutterEngine(@NonNull flutterEngine: FlutterEngine) {
        GeneratedPluginRegistrant.registerWith(flutterEngine);
    }*/
}
```

**Comment by mklim** [comment]
Thanks @woprandi for filing the issue and diagnosing this! Just as a warning, that workaround should fix this specific bug for `firebase_messaging` but may cause issues in other plugins like `google_maps_flutter` because of a bug with the current registration flow. We're tracking the larger issue at flutter/flutter#49365.

**Comment by woprandi** [comment]
@mklim Thanks for the info, I wasn't able to find the goal of the MainActivity content.

**Comment by hp1227** [comment]
@mklim is there any updates?

**Comment by woprandi** [comment]
Thanks for your work @mklim

**Comment by davidwilcox** [comment]
I just tried against master (looks like the commit is not yet merged to stable), and it worked for me!

**Comment by rodrigolmacedo** [comment]
> 
> 
> I just tried against master (looks like the commit is not yet merged to stable), and it worked for me!

here too, but with backgroundmessage register bug.

**Comment by viniciusroland** [comment]
woprandi comment solves the issue for android, but how to solve the same issue for iOS?

**Comment by noobcoder17** [comment]
> A working workaround :
> 
> ```
> class MainActivity: FlutterActivity() {
>    /*  override fun configureFlutterEngine(@NonNull flutterEngine: FlutterEngine) {
>         GeneratedPluginRegistrant.registerWith(flutterEngine);
>     }*/
> }
> ```

This works for me in Android.

**Comment by kbrmimbyl** [comment]
We're on Flutter 1.12.13+hotfix.7 and facing this issue right now. Is this bug fixed in 1.12.13+hotfix.8?

**Comment by DenisBarzanov** [comment]
> We're on Flutter 1.12.13+hotfix.7 and facing this issue right now. Is this bug fixed in 1.12.13+hotfix.8?

I am running 1.12.13+hotfix.8 on stable and it still does not work. Installing [fvm](https://github.com/leoafarias/fvm) to switch between master and stable.

**Comment by fxsalazar** [comment]
I'm running `v1.12.13+hotfix.8` but with `firebase_messaging: ^5.1.6` because when trying to update to `v6.x` the notifications on iOS stoped working. Not really sure where or what has the issue.

**Comment by pdelgadohurtado** [comment]
Still not working in `v1.12.13+hotfix.8` with `firebase_messaging: ^6.0.12`

**Comment by nizarhmain** [comment]
Tried it with ```flutter v1.15.19 dev```. It seems to work for me 
https://storage.googleapis.com/flutter_infra/releases/dev/windows/flutter_windows_v1.15.19-dev.zip

**Comment by DenisBarzanov** [comment]
Yes, the master branch seems fine

**Comment by schankam** [comment]
When will this be merged to Stable ? Any updates ? As of today, the problem still persist.

**Comment by woutergoossens** [comment]
When will this be in the stable version?

**Comment by p30arena** [comment]
firebase_messaging 6.0.13
Flutter (Channel master, v1.16.4-pre.52, on Microsoft Windows [Version 6.3.9600], locale en-US)

**Comment by lucas-sesti** [comment]
@p30arena the version 6.0.13 is correct ? i'm using and still calling twice in the onResume and others...

**Comment by p30arena** [comment]
@lucas-sesti 

maybe you should check your server logic,
I've tested it in two production apps on Android and iOS.

**Comment by lucas-sesti** [comment]
But in not stable channel of flutter alright? I'm using the Stable channel, now i see you said:
"Flutter (Channel master, v1.16.4-pre.52, on Microsoft Windows [Version 6.3.9600], locale en-US)"...

**Comment by lexyfeito** [comment]
Hi, any ideas when this will be merge to stable?

**Comment by theochampion** [comment]
Any ETA on the merge to stable? :)

**Comment by Danaru87** [comment]
Not resolved with flutter 1.12.13+hotfix.9.

Any news about merging into stable version ?

**Comment by vinkalathmin** [comment]
This is very serious issue. Is there there any work around, we have an urgent release.
Please help

**Comment by braindamagedman** [comment]
Hello guys. I think, after flutter upgrade to 1.12, you need to migrate your projects following this guide: https://github.com/flutter/flutter/wiki/Upgrading-pre-1.12-Android-projects#full-flutter-app-migration

> body of the MainActivity is now empty

> The new FlutterActivity no longer requires manually registering your plugins. It will now perform the registration automatically when the underlaying FlutterEngine is created.

**Comment by huulbaek** [comment]
I'm doing something like this:

```dart
if ((DateTime.now().microsecondsSinceEpoch - _lastOnResumeCall) > 1000000) {
  _lastOnResumeCall = DateTime.now().microsecondsSinceEpoch;
  _navigateToItemDetail(message);
}
```

Essentially testing if the callback has just been called.

**Comment by mtannerfors** [comment]
Fixed and working well on flutter beta. Thanks @mklim!

**Comment by MichaelOttink** [comment]
Any ETA on a fix for this? Still happening on the latest flutter version.

**Comment by Akora-IngDKB** [comment]
> Fixed and working well on flutter beta. Thanks @mklim!

I'm on beta. Still facing the same issue.

**Comment by Danaru87** [comment]
> Hello guys. I think, after flutter upgrade to 1.12, you need to migrate your projects following this guide: https://github.com/flutter/flutter/wiki/Upgrading-pre-1.12-Android-projects#full-flutter-app-migration
> 
> > body of the MainActivity is now empty
> 
> > The new FlutterActivity no longer requires manually registering your plugins. It will now perform the registration automatically when the underlaying FlutterEngine is created.

Following migration guide resolved this problem for me ;) Thx !

**Comment by om-ha** [comment]
Got this issue on iOS too!

# Version
firebase_crashlytics: ^0.1.3+3

# Full Log
```
Xcode build done.                                           70.7s
Configuring the default Firebase app...
[Crashlytics] Version 3.14.0 (144)
Configured the default Firebase app __FIRAPP_DEFAULT.
**[Fabric] [Fabric +with] called multiple times. Only the first call is honored, please pass all kits you wish to initialize**
[GoogleDataTransport][I-GDTCOR001006] (/Users/xxxxxxxxxx/ios/Pods/GoogleDataTransport/GoogleDataTransport/GDTCORLibrary/GDTCOREvent.m:59) : There was an error saving the new counter value to disk.
[GoogleDataTransport][I-GDTCOR001006] (/Users/xxxxxxxxxx/ios/Pods/GoogleDataTransport/GoogleDataTransport/GDTCORLibrary/GDTCOREvent.m:59) : There was an error saving the new counter value to disk.
WARNING: OneSignal has detected that your application delegate implements a deprecated method (application:didReceiveLocalNotification:). Please note that this method has been officially deprecated and the OneSignal SDK will no longer call it. You should use UNUserNotificationCenter instead
WARNING: OneSignal has detected that your application delegate implements a deprecated method (application:didReceiveLocalNotification:). Please note that this method has been officially deprecated and the OneSignal SDK will no longer call it. You should use UNUserNotificationCenter instead
```

# Flutter Doctor
```
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, v1.12.13+hotfix.9, on Mac OS X 10.14.5 18F132,
    locale en-US)
 
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.3)
[✓] Xcode - develop for iOS and macOS (Xcode 11.3)
[!] Android Studio (version 3.6)
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
[✓] VS Code (version 1.44.2)
[✓] Connected device (1 available)

! Doctor found issues in 1 category.
```

**Comment by om-ha** [comment]
``` dart
void main() async {
  WidgetsFlutterBinding.ensureInitialized();

  // setup OneSignal
  await AppNotificationsManager.initializeOneSignal();

  // setup Crashlytics.
  setUpCrashlytics();

  // initialize bloc injector
  var container =
      await BlocInjector.create(BlocModule(), CoreModule(), ScreensModule());

  runZoned(() {
    runApp(container.app);
  }, onError: Crashlytics.instance.recordError);
}

void setUpCrashlytics() {
  // Set `enableInDevMode` to true to see reports while in debug mode
  // This is only to be used for confirming that reports are being
  // submitted as expected. It is not intended to be used for everyday
  // development.
  Crashlytics.instance.enableInDevMode = false;

  // Pass all uncaught errors from the framework to Crashlytics.
  FlutterError.onError = Crashlytics.instance.recordFlutterError;
}
```

**Comment by m4ff** [comment]
Flutter 1.7 was released to the stable channel, which should include a fix to this bug.

**Comment by om-ha** [comment]
Wow you're actually correct.. `Flutter 1.17.0` was released to stable channel. I just updated to latest hotfix 2 days ago!

**Comment by mklim** [comment]
The Android issue originally mentioned here has been fixed on all Flutter channels at this point, so I'm going to close this as fixed.

From some of the comments I suspect that there may have been multiple issues reported on this bug. If you're on the latest stable version of Flutter (currently `v1.17.0`) and still seeing this behavior, please [file a new issue](https://github.com/FirebaseExtended/flutterfire/issues/new/choose).

**Comment by MinaSamirCap** [comment]
> A working workaround :
> 
> ```
> class MainActivity: FlutterActivity() {
>    /*  override fun configureFlutterEngine(@NonNull flutterEngine: FlutterEngine) {
>         GeneratedPluginRegistrant.registerWith(flutterEngine);
>     }*/
> }
> ```

this solution was so great ... but it makes some issues with other packages like image picker

the wrong with my code was that I was calling the firebase init method in the build method
of the application and that was the cause of duplication because the build method maybe
called more than one time.
the solution for me was to call the firebase init method in the initState() method of Widget that will 
enable init of firebase to be called only onetime and that prevent the duplication.

that works for me I hope for all too

**Comment by Chethansmash7** [comment]
> > A working workaround :
> > ```
> > class MainActivity: FlutterActivity() {
> >    /*  override fun configureFlutterEngine(@NonNull flutterEngine: FlutterEngine) {
> >         GeneratedPluginRegistrant.registerWith(flutterEngine);
> >     }*/
> > }
> > ```
> 
> this solution was so great ... but it makes some issues with other packages like image picker
> 
> the wrong with my code was that I was calling the firebase init method in the build method
> of the application and that was the cause of duplication because the build method maybe
> called more than one time.
> the solution for me was to call the firebase init method in the initState() method of Widget that will
> enable init of firebase to be called only onetime and that prevent the duplication.
> 
> that works for me I hope for all too

Could you explain it with your code,because your explanatin is kind of confusing me...

**Comment by MinaSamirCap** [comment]
What I want to say it that
You do not need to comment this code in the MainActiviy
class MainActivity: FlutterActivity() {
   /*  override fun configureFlutterEngine(@NonNull flutterEngine: FlutterEngine) {
        GeneratedPluginRegistrant.registerWith(flutterEngine);
    }*/
}

just initialize the firebase in the initState() function instead of build() function
that prevented the duplicated notification for me

**Comment by ratnapriya4g** [comment]
**This is the perfect code for this issue.
Just we need to do is to declare a static bool variable, It will work with onResume().**
**Before only some logic is missing, that's why it is calling twice.**

`static bool isNotified=false;`
` onMessage:
          // ignore: missing_return
          (Map<String, dynamic> msg) {
            if(!isNotified)
            {
              print("onMessage called");
              print(msg);
              DocumentReference documentReference =
              Firestore.instance.collection('PushNotifications').document();
              documentReference.setData({
                "payload": msg
              });
      }
            isNotified = !isNotified;
    }`

@samyu99
@woprandi  
![picturemessage_f0rgyg5l fgp](https://user-images.githubusercontent.com/38585678/86094972-d9e64e80-bace-11ea-8599-0a17a69dff84.png)

**Comment by pdivita** [comment]
I can confirm that this issue is fixed if you create a new project using Flutter version 1.17.4.
For the record I'm using firebase_messaging: ^6.0.9

**Comment by kharisazhar** [comment]
> I can confirm that this issue is fixed if you create a new project using Flutter version 1.17.4.
> For the record I'm using firebase_messaging: ^6.0.9

still same getting bug

**Comment by samyu99** [comment]
Use bool  variable ,it is working perfect for me.

`static bool isNotified=false;
onMessage: // ignore: missing_return (Map<String, dynamic> msg) { if(!isNotified) { print("onMessage called"); print(msg); DocumentReference documentReference = Firestore.instance.collection('PushNotifications').document(); documentReference.setData({ "payload": msg }); } isNotified = !isNotified; }
`
@kharisazhar

**Comment by cayau-wamsa** [comment]
For me, works on update Flutter version 1.17.5 and firebase_messaging: ^6.0.16

**Comment by fnicastri** [comment]
I'm facing the same issue on iOS

**Comment by rivetingpeppermint** [comment]
Still facing the same issue on both Android and iOS. I'm using `firebase_messaging: ^6.0.16` and Flutter 1.17.5.

**Comment by Chethansmash7** [comment]
I think I solved the issue.But I need to be sure as to what steps you are following.Those who have the issue please reply to this comment with the steps by step procedure...

**Comment by fnicastri** [comment]
@Chethansmash7 

1. Clone this repo
2. change app id and put the new GoogleService-Info.plist in the project
3. build packages/firebase_messaging/example
4. send a notification while the app is in foreground
5. onMessage get called twice
```
flutter: onMessage: {body: Body, google.c.sender.id: 276852428296, google.c.a.e: 1, click_action: FLUTTER_NOTIFICATION_CLICK, aps: {alert: {title: Title, body: Test notification}, content-available: 1}, title: title, gcm.message_id: 1594807744573443, priority: high, content_available: true, gcm.notification.priority: high}

flutter: onMessage: {click_action: FLUTTER_NOTIFICATION_CLICK, google.c.sender.id: 276852428296, google.c.a.e: 1, content_available: true, aps: {alert: {title: Title, body: Test notification}, content-available: 1}, title: title, gcm.message_id: 1594807744573443, body: Body, priority: high, gcm.notification.priority: high}
```

Flutter doctor
```
[✓] Flutter (Channel stable, v1.17.5, on Mac OS X 10.15.5 19F101, locale en-US)
    • Flutter version 1.17.5 at /Users/cico/src/flutter
    • Framework revision 8af6b2f038 (2 weeks ago), 2020-06-30 12:53:55 -0700
    • Engine revision ee76268252
    • Dart version 2.8.4

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.0)
    • Android SDK at /Users/cico/Library/Android/sdk
    • Platform android-30, build-tools 30.0.0
    • ANDROID_SDK_ROOT = /Users/cico/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.5)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.5, Build version 11E608c
    • CocoaPods version 1.9.3

[✓] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 47.1.2
    • Dart plugin version 193.7361
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)

```

Can you share your solution?

**Comment by fnicastri** [comment]
I've noticed that the callback onMessage got called twice in FLTFirebaseMessagingPlugin.m

commenting out 
```
if #available(iOS 10.0, *) {
  UNUserNotificationCenter.current().delegate = self as? UNUserNotificationCenterDelegate
}
```
did the trick.
But obviously it is just a work around, I'm not aware of the possible repercussions.

The flow is:
1. `(BOOL)application:(UIApplication *)application
    didReceiveRemoteNotification:(NSDictionary *)userInfo
          fetchCompletionHandler:(void (^)(UIBackgroundFetchResult result))completionHandler` 
get called and
2. it calls `void)didReceiveRemoteNotification:(NSDictionary *)userInfo`
3. `- (void)userNotificationCenter:(UNUserNotificationCenter *)center
       willPresentNotification:(UNNotification *)notification
         withCompletionHandler:(void (^)(UNNotificationPresentationOptions))completionHandler
    NS_AVAILABLE_IOS(10.0) ` get called
4. and it call a second time `[_channel invokeMethod:@"onMessage" arguments:userInfo]`

**Comment by rivetingpeppermint** [comment]
I tried @fnicastri's workaround and it works for onMessage, but onResume is still called twice on iOS. onResume is called when the app is _still in the background_ as well as when the app is brought to the foreground. Is it supposed to be this way? From reading the documentation, I thought onResume will only be called when the app is brought to the foreground, and not when it's in the background too.

**Comment by fnicastri** [comment]
@rivetingpeppermint @TahaTesser 
Yes, digging more I noticed the same issue.
This is a serious issue that makes notifications unusable.

**Comment by rivetingpeppermint** [comment]
I also just found out that when I run on iOS, onResume and onLaunch are called when we open the app normally (not from clicking the notification in the system tray). This behavior doesn't happen with Android. I used the iPad Pro (9.7-inch, running iOS 13.5.1) and iPhone 7 (running iOS 12.4.1).

**Comment by fnicastri** [comment]
@TahaTesser 
Thanks for reopening this bug.
you could change the platform tag too and maybe the title.

**Comment by hasanzia1993** [comment]
> I confirm it's only on Android

it on iOS too

**Comment by mktdevcwi** [comment]
+1

**Comment by HUYTHAINGUYEN** [comment]
Does anyone have a workaround for this problem yet?

**Comment by eyoeldefare** [comment]
> Does anyone have a workaround for this problem yet?

init firebase only once, it worked for me

**Comment by fnicastri** [comment]
> > Does anyone have a workaround for this problem yet?
> 
> init firebase only once, it worked for me

Can you explain it?

**Comment by ollydixon** [comment]
Also seeing the issue but only on devices (iOS). On the emulator and simulator both Android and iOS it's once. This is a major issue for us since all of our apps rely on FCM and certain actions are now appearing twice in the app for dialogs etc and that is really s**t.

**Comment by ollydixon** [comment]
It's fixed with the new version. I'm using this current configuration:

```
  firebase_core: 0.5.0
  firebase_auth: 0.18.0
  firebase_messaging: 7.0.0
```

**Comment by rivetingpeppermint** [comment]
@ollydixon what additional setup?

**Comment by ollydixon** [comment]
@rivetingpeppermint ahh sorry, it's on the github/pub page.

**Comment by fnicastri** [comment]
@ollydixon @TahaTesser

> It's fixed with the new version. I'm using this current configuration: (Don't use carets it's not good)
> 
> ```
>   firebase_core: 0.5.0
>   firebase_auth: 0.18.0
>   firebase_messaging: 7.0.0
> ```
> 
> Be aware there is now some additional Android setup 👎

It's not.
I still get onResume called twice on iOS. On Android it's fine now (no additional setup needed)
And if I add 
```
if (@available(iOS 10.0, *)) {
  [UNUserNotificationCenter currentNotificationCenter].delegate = (id<UNUserNotificationCenterDelegate>) self;
}
```
I get nothing at all.

**Comment by dppatel03** [comment]
firebase_core: 0.5.0
  firebase_auth: 0.18.0
  firebase_messaging: 7.0.0

I am getting same issue at iOS.

**Comment by kartercs** [comment]
Having the same issue on IOS for onMessage and onResume. 
If it can help the second message object have the same informations as the first one but in different order like this : 
>{sender: ***, google.c.sender.id: ***, google.c.a.e: ***, click_action: FLUTTER_NOTIFICATION_CLICK, aps: {alert: {title: ***, body: ***}, sound: ***, content-available: ***}, sound: ***, content_available: ***, priority: ***, gcm.message_id: ***, gcm.notification.priority: ***}

>{click_action: FLUTTER_NOTIFICATION_CLICK, sender: ***, google.c.a.e: ***, google.c.sender.id: ***, aps: {alert: {title: ***e, body: ***}, sound: default, content-available: 1}, sound: ***, gcm.message_id: ***, priority: ***, content_available: ***, gcm.notification.priority: ***}

**Comment by salva73** [comment]
Same issue here, using flutter v. 1.22.0 and firebase messaging 7.0.3

**Comment by TahaTesser** [comment]
@salva73
Could you  please file a new issue with all the steps
